File: backend/src/main/java/com/centerport/config/security/Roles.java
package com.centerport.config.security;

/**
 * Central definition of the Keycloak realm role names used for authorization.
 *
 * <p>These constants hold the <b>bare</b> role name (no {@code ROLE_} prefix),
 * matching the value that appears in the Keycloak token's
 * {@code realm_access.roles} array. They are intended for use with Spring
 * Security's {@code hasRole(..)} / {@code hasAnyRole(..)} expressions and the
 * URL rules in {@code SecurityConfig}, both of which add the {@code ROLE_}
 * prefix themselves.
 *
 * <p>In {@code @PreAuthorize} annotations use them as, e.g.:
 * <pre>{@code @PreAuthorize("hasAnyRole('" + Roles.ADMIN + "','" + Roles.LABORATORY + "')")}</pre>
 * or simply the literal string {@code @PreAuthorize("hasAnyRole('ADMIN','LABORATORY')")}.
 */
public final class Roles {

    private Roles() {
    }

    public static final String ADMIN = "ADMIN";
    public static final String INFORMATION = "INFORMATION";
    public static final String PSYCHOLOGY = "PSYCHOLOGY";
    public static final String LABORATORY = "LABORATORY";
    public static final String RELEASING = "RELEASING";
    public static final String ACCOUNTING = "ACCOUNTING";
}

File: frontend/src/lib/roles.ts
/**
 * Application roles and the module → role access map.
 *
 * This MIRRORS the authoritative rules enforced by the Spring Boot backend
 * (see backend/documentation/keycloak-security.md). The frontend uses this only
 * for UX — hiding nav items and guarding routes the user cannot use. The
 * backend still returns 403 if a user calls a forbidden endpoint, so these
 * checks can never be relied on for security.
 */

/** Keycloak realm roles (bare names, matching `realm_access.roles`). */
export type Role =
  | "ADMIN"
  | "INFORMATION"
  | "PSYCHOLOGY"
  | "LABORATORY"
  | "RELEASING"
  | "ACCOUNTING";

/**
 * Maps a top-level route (by its `href` in navigation.ts) to the roles that may
 * access it. `ADMIN` is added to every entry automatically by
 * {@link canAccessRoute}, so it is omitted here.
 */
const ROUTE_ROLES: Record<string, Role[]> = {
  "/dashboard": ["INFORMATION", "PSYCHOLOGY", "LABORATORY", "RELEASING"],
  "/visit": ["INFORMATION"],
  "/profile": ["INFORMATION", "RELEASING"],
  "/seabase": ["RELEASING"],
  "/laboratory": ["LABORATORY"],
  "/mlc": ["RELEASING"],
  "/panama": ["RELEASING"],
  "/landbase": ["RELEASING"],
  "/psychology": ["PSYCHOLOGY"],
  // Accounting now lives entirely in the /sale POS workspace (billing endpoints
  // /api/receivables, /api/items, /api/customers, /api/products are restricted
  // to ADMIN + ACCOUNTING by the backend). There are no standalone accounting
  // routes to gate here anymore.
  // Super Admin only: an empty required-role list means no non-ADMIN role
  // grants access, so only the ADMIN short-circuit in canAccessRoute passes.
  "/medical-personnel": [],
};

/** True if any of the user's roles grant access to the given route href. */
export function canAccessRoute(href: string, roles: readonly string[]): boolean {
  if (roles.includes("ADMIN")) {
    return true;
  }
  // Match the longest known route prefix (e.g. /profile/123 → /profile).
  const key = Object.keys(ROUTE_ROLES)
    .filter((route) => href === route || href.startsWith(route + "/"))
    .sort((a, b) => b.length - a.length)[0];

  if (!key) {
    // Unlisted routes are visible to any authenticated user.
    return true;
  }
  return ROUTE_ROLES[key].some((role) => roles.includes(role));
}

/**
 * The route a user should land on after login.
 *
 * Everyone is sent to `/dashboard` by default (see `app/page.tsx`), but some
 * roles cannot access it — the backend restricts `/api/dashboard/**` to
 * ADMIN, INFORMATION, PSYCHOLOGY, LABORATORY and RELEASING, so a dashboard
 * landing for anyone else (currently ACCOUNTING) only produces 403s. Those
 * users are routed to their primary workspace instead.
 *
 * ADMIN always keeps the dashboard because ADMIN can access every route. The
 * check is expressed in terms of `canAccessRoute` so it stays in lockstep with
 * the module access map above.
 */
export function landingRoute(roles: readonly string[]): string {
  if (canAccessRoute("/dashboard", roles)) {
    return "/dashboard";
  }
  // ACCOUNTING (and any future non-dashboard role) starts in the POS/sale
  // workspace, which is open to any authenticated user.
  return "/sale";
}

/** True if the user holds at least one of the required roles (ADMIN always passes). */
export function hasAnyRole(
  roles: readonly string[],
  required: readonly Role[]
): boolean {
  if (roles.includes("ADMIN")) {
    return true;
  }
  return required.some((role) => roles.includes(role));
}
